torch/csrc/jit/codegen/cuda/kernel.h

```diff
@@ -74,25 +84,40 @@ class TORCH_CUDA_API Kernel final : public NonCopyable {
     return outputs_;
   }
 
-  const auto& exprs() const {
-    return exprs_;
+  const auto& topLevelExprs() const {
+    return top_level_exprs_;
   }
 
   const KernelSummary& summary() const {
     return summary_;
   }
 
   const ThreadPredicateMap& predicateMap() const {
-    return predicate_map_;
+    return *predicate_map_;
+  }
+
+  //! Register a new Kernel IR node
+  //!
+  //! \note This is a specialized helper for kir::IrBuilder, not
+  //!   intendted for general use
+  //!
+  void registerIrNode(std::unique_ptr<Statement> node) {
+    ir_nodes_.push_back(std::move(node));
   }
 
  private:
   // Analyze the kernel IR and caches the summary of interesting data
   void analyze();
 
  private:
-  // Lowered expressions
-  std::vector<Expr*> exprs_;
+  // Kernel IR nodes
+  std::vector<std::unique_ptr<Statement>> ir_nodes_;
+
+  // Map from value to its definition expression
+  std::unordered_map<const Val*, Expr*> definitions_;
+
+  // Top level expressions
+  std::vector<Expr*> top_level_exprs_;
```

Comment:
What do you mean by `top level expressions`?

Reply:
This is the list of expressions at the top kernel scope, not all the expressions in the Kernel IR. These top level expressions may have arbitrary nested expressions (ex. for loops) - ie. the reason we need the recursive traversal in `LocalSyncInserter`.

Reply:
Makes sense, since `exprs` implies all expressions in the kernel. 👍